[1.21.4] Encourage more usage of IConditionBuider

ApexModder opened the pull request:
This PR aims to encourage modders to make more use of `IConditionBuilder` rather than constructing condition instances themselves

this is done by
- patching `DataProvider`s which support conditions to now implement `IConditionBuilder` making all of its helpers freely available to all inheritors.
- adding a static `of` method allowing users to construct an instance in places where they do not have a `DataProvider` or can implement it themselves.
- deprecating and moving all shortcut constructors into `ICondtionBuilder`
- deprecating the old `TRUE`/`FALSE` conditions in favor of the new `always`/`never` conditions

ApexModder committed:
Implement IConditionBuilder onto DataProvider

ApexModder committed:
Deprecate `TRUE` and `FALSE` conditions, renamed to `always` and `never`

ApexModder committed:
Encourage usage of `IConditionBuilder` more

by deprecating all shortcut constructors and moving them into `IConditionBuilder`

ApexModder committed:
Convert `ItemExistsCondition` back into a nonfinal class

less of a breaking change

Technici4n commented on patches/net/minecraft/data/DataProvider.java.patch:
```diff
@@ -1,9 +1,11 @@
  import org.slf4j.Logger;
  
- public interface DataProvider {
+-public interface DataProvider {
```
Do we really want it in every data provider? Or should we be more selective and only implement it where we know it is usable?

ApexModder replied:
I remembered seeing this PR (#1315) and thought our condition system supported any data file now, but guess I my assumption on that was wrong as it was closed not merged.

I can go through and find all the providers which directly support conditions, rather than implementing onto `DataProvider`.

Technici4n commented on src/main/java/net/neoforged/neoforge/common/conditions/TagEmptyCondition.java:
```diff
@@ -19,14 +19,26 @@ public record TagEmptyCondition(TagKey<Item> tag) implements ICondition {
+    /**
+     * @deprecated Use {@link IConditionBuilder}
+     */
+    @Deprecated(forRemoval = true, since = "1.12.3")
```
1.12.3?

ApexModder replied:
Good catch, should be fixed now

Technici4n commented on tests/src/main/java/net/neoforged/neoforge/debug/data/registries/DatapackEntryTests.java:
```diff
@@ -46,8 +45,8 @@ static void conditionalDatapackEntries(final DynamicTest test, final Registratio
                 conditions -> {
-                    conditions.accept(CONDITIONAL_FALSE_DAMAGE_TYPE, FalseCondition.INSTANCE);
-                    conditions.accept(CONDITIONAL_TRUE_DAMAGE_TYPE, TrueCondition.INSTANCE);
+                    conditions.accept(CONDITIONAL_FALSE_DAMAGE_TYPE, IConditionBuilder.of().never());
```
Makes me wonder if static methods in the builder would work instead. 🤔

ApexModder replied:
I was thinking that too, but that would have to wait until 21.4
Since afaik cant make static versions cause the method signature will conflict the the default ones.

Technici4n commented on src/main/java/net/neoforged/neoforge/common/conditions/IConditionBuilder.java:
```diff
@@ -17,14 +19,32 @@ default ICondition and(ICondition... values) {
 
-    default ICondition FALSE() {
+    default ICondition never() {
+        // TODO: Maybe also rename class and registry entry to 'never' as well
```
This kind of TODO will get forgotten and come back to haunt us in a few years. :P

ApexModder replied:
Deprecated the old `TRUE`/`FALSE` in favor of newly added `always`/`never` conditions

ApexModder committed:
Rename and deprecate `TRUE`/`FALSE` conditions in favour of `always`/`never`

ApexModder committed:
Fix typo in deprecation markers

ApexModder committed:
Implement `IConditionBuilder` onto providers which support the conditional system

ApexModder committed:
Remove deprecated elements

Convert `ItemExistsCondition` to a record

Technici4n commented on patches/net/minecraft/data/recipes/RecipeProvider.java.patch:
```diff
@@ -10,6 +19,15 @@
+     }
+ 
+-    public abstract static class Runner implements DataProvider {
++    public abstract static class Runner implements DataProvider, net.neoforged.neoforge.common.conditions.IConditionBuilder {
```
I don't think it makes sense to implement this on `Runner`?

Technici4n commented on src/main/java/net/neoforged/neoforge/common/NeoForgeMod.java:
```diff
@@ -357,13 +357,14 @@ public class NeoForgeMod {
     public static final DeferredHolder<MapCodec<? extends ICondition>, MapCodec<OrCondition>> OR_CONDITION = CONDITION_CODECS.register("or", () -> OrCondition.CODEC);
     public static final DeferredHolder<MapCodec<? extends ICondition>, MapCodec<TagEmptyCondition>> TAG_EMPTY_CONDITION = CONDITION_CODECS.register("tag_empty", () -> TagEmptyCondition.CODEC);
-    public static final DeferredHolder<MapCodec<? extends ICondition>, MapCodec<TrueCondition>> TRUE_CONDITION = CONDITION_CODECS.register("true", () -> TrueCondition.CODEC);
+
```
Extra space got added here.

Technici4n commented on src/main/java/net/neoforged/neoforge/common/conditions/IConditionBuilder.java:
```diff
@@ -33,16 +35,36 @@ default ICondition or(ICondition... values) {
+        return tagEmpty(ResourceLocation.fromNamespaceAndPath(namespace, tagPath));
+    }
+
+    default ICondition tagEmpty(String tagPath) {
```
`path` is misleading since it also includes the namespace. I'd probably remove the `String` methods, that's an outdated pattern. :wink:

Technici4n commented on src/main/java/net/neoforged/neoforge/common/conditions/IConditionBuilder.java:
```diff
@@ -33,16 +35,36 @@ default ICondition or(ICondition... values) {
+        return new ItemExistsCondition(ResourceLocation.fromNamespaceAndPath(namespace, path));
+    }
+
+    default ICondition itemExists(String itemName) {
```
Remove this one? It's confusing I think, RL ftw!

ApexModder replied:
tbh ive been looking at this and asking myself why its strictly limited to items only?
conditions have a full registry access, could we expand this and tagEmpty to support any registry type?

say a mod wanted to do something if a enchantment exists, or block tag is empty. why limit them to strictly items?

Technici4n commented on src/main/java/net/neoforged/neoforge/common/conditions/IConditionBuilder.java:
```diff
@@ -59,4 +81,8 @@ default ICondition featureFlagsEnabled(FeatureFlag... requiredFlags) {
         }
     }
+
+    static IConditionBuilder of() {
```
I'd remove this, it's a bit weird?

Technici4n commented on src/main/java/net/neoforged/neoforge/common/conditions/IConditionBuilder.java:
```diff
@@ -17,12 +19,12 @@ default ICondition and(ICondition... values) {
         return new AndCondition(List.of(values));
     }
```
How about we make all these methods static and turn this class into a helper class? E.g. `NeoForgeConditions.never()`. That would match what we do in other places better.

Technici4n commented on tests/src/main/java/net/neoforged/neoforge/oldtest/conditions/ConditionalCodecTest.java:
```diff
@@ -141,7 +141,7 @@ public static void testWriteWithConditions() {
                       ],
```
Uses of true/false in this file need renaming.

Technici4n reviewed (changes requested):
Also: did you check for `true` and `false` conditions in the non-generated resource files?

ApexModder committed:
Convert helpers to being static

Rename `IConditionBuilder` -> `NeoForgeConditions`

ApexModder committed:
Rename legacy `true`/`false` condition references

ApexModder committed:
Remove whitespace in `NeoForgeMod`

ApexModder committed:
Fix misplaced brackets during rebase

ApexModder committed:
Remove deprecated constructor from `ConditionContext`

ApexModder committed:
Expand `ItemExists` and `TagEmpty` conditions to support any registry type

Defaulting to `ITEM` registry if omitted
Rename `ItemExistsCondition` to `ElementExistsCondition`

Technici4n commented on src/main/java/net/neoforged/neoforge/common/conditions/ElementExistsCondition.java:
```diff
@@ -0,0 +1,34 @@
+import net.minecraft.resources.ResourceKey;
+import net.minecraft.resources.ResourceLocation;
+
+public record ElementExistsCondition<TRegistry>(ResourceKey<TRegistry> registryKey) implements ICondition {
```
```suggestion
public record ElementExistsCondition<T extends Registry<?>>(ResourceKey<T> registryKey) implements ICondition {
```
does this work? Not a fan of `TSomething`.

ApexModder replied:
I could change it to `T` but having it extend `Registry` doesnt help as we are not looking for a "Registry" instance just a "registry type". That being `Item`, `Block`, `BlockEntityType`, `EntityType` etc, none of which extend `Registry`.

Edit: another thing about the type being bound to `Registry`, the condition would be checking if the registry itself exists, not some entry within the registry.

Technici4n replied:
Ok, no need for `extends Registry<?>` then. I was confused by the `TRegistry` naming.

Technici4n commented on src/main/java/net/neoforged/neoforge/common/conditions/ElementExistsCondition.java:
```diff
@@ -0,0 +1,34 @@
+
+public record ElementExistsCondition<TRegistry>(ResourceKey<TRegistry> registryKey) implements ICondition {
+    public static final MapCodec<ElementExistsCondition<?>> CODEC = RecordCodecBuilder.mapCodec(instance -> instance
+            .group(ResourceLocation.CODEC.lenientOptionalFieldOf("registry", Registries.ITEM.location()).forGetter(condition -> condition.registryKey().registry()),
```
Don't use lenient optional, it might swallow errors. I wonder if `ResourceKey.codec(Registries.ROOT_REGISTRY_NAME)` could be used here.

ApexModder replied:
`ResourceKey.codec` wants a `ResourceKey<? extends Registry<T>>`, `ResourceKey.ROOT_REGISTRY_NAME` is a `ResourceLocation`.
Dont think i can use that method due to it being typed and the registries resource key.

The way i have also designed the codec here means existing json files outside of renaming the `item` value to `element` should just work.

Technici4n replied:
Yes it's good then, just replace the lenient optional by a normal optional. Lenient optional will cause nonsensical values such as `{}` to be silently ignored.

Technici4n commented on src/main/java/net/neoforged/neoforge/common/conditions/ElementExistsCondition.java:
```diff
@@ -0,0 +1,34 @@
+import net.minecraft.resources.ResourceKey;
+import net.minecraft.resources.ResourceLocation;
+
+public record ElementExistsCondition<TRegistry>(ResourceKey<TRegistry> registryKey) implements ICondition {
```
I don't like the name "element" which is not used at all in other contexts referring to registry entries.

Technici4n replied:
Idea: `RegisteredCondition`? And replace `element` by `value`? 🤔

Technici4n commented on src/main/java/net/neoforged/neoforge/common/conditions/NeoForgeConditions.java:
```diff
@@ -0,0 +1,101 @@
+import net.minecraft.world.flag.FeatureFlagSet;
+import org.apache.commons.lang3.ArrayUtils;
+
+public interface NeoForgeConditions {
```
Make it a `public final class`, no?

ApexModder replied:
Could do that i just personally dont see the point, extending this class gives you no special benefits or methods, and being an interface lessens the amount of `public static [final]` needed everywhere as everything by default is `public static [final]`.

If preferred though i can change it to be a final class, i just personally dont see any reason for it to be a final class

Technici4n replied:
A `public final class` with a `private` constructor and only static methods conveys the intent that it's a helper class. An interface conveys the intent that it can be implemented. We should convey the right intent as that makes our code more readable and understandable.

Technici4n commented on src/main/java/net/neoforged/neoforge/common/conditions/TagEmptyCondition.java:
```diff
@@ -8,27 +8,18 @@
-                    .group(
-                            ResourceLocation.CODEC.xmap(loc -> TagKey.create(Registries.ITEM, loc), TagKey::location).fieldOf("tag").forGetter(TagEmptyCondition::tag))
-                    .apply(builder, TagEmptyCondition::new));
+public record TagEmptyCondition<TRegistry>(TagKey<TRegistry> tag) implements ICondition {
```
Not a fan of `TRegistry` compared to `T`.

Technici4n commented on src/main/java/net/neoforged/neoforge/common/conditions/TagEmptyCondition.java:
```diff
@@ -8,27 +8,18 @@
-                    .apply(builder, TagEmptyCondition::new));
+public record TagEmptyCondition<TRegistry>(TagKey<TRegistry> tag) implements ICondition {
+    public static final MapCodec<TagEmptyCondition<?>> CODEC = RecordCodecBuilder.mapCodec(instance -> instance
+            .group(ResourceLocation.CODEC.lenientOptionalFieldOf("registry", Registries.ITEM.location()).forGetter(condition -> condition.tag().registry().location()),
```
Don't use lenient optional.

Technici4n commented on tests/src/main/java/net/neoforged/neoforge/oldtest/DataGeneratorTest.java:
```diff
@@ -168,8 +168,8 @@ public static void gatherData(GatherDataEvent.Client event) {
                         new WithConditions<>(new OverlayMetadataSection.OverlayEntry(new InclusiveRange<>(0, Integer.MAX_VALUE), "pack_overlays_test")),
-                        new WithConditions<>(new OverlayMetadataSection.OverlayEntry(new InclusiveRange<>(0, Integer.MAX_VALUE), "conditional_overlays_enabled"), new ModLoadedCondition("neoforge")),
-                        new WithConditions<>(new OverlayMetadataSection.OverlayEntry(new InclusiveRange<>(0, Integer.MAX_VALUE), "conditional_overlays_enabled"), new ModLoadedCondition("does_not_exist")))))
+                        new WithConditions<>(new OverlayMetadataSection.OverlayEntry(new InclusiveRange<>(0, Integer.MAX_VALUE), "conditional_overlays_enabled"), NeoForgeConditions.modLoaded(NeoForgeVersion.MOD_ID)),
```
Note that `NeoForgeVersion` is internal and should ideally not be used outside of NeoForge. (At the moment at least).

ApexModder replied:
Sure but this is a internal test class, i see no problem using internal things here.
but if using internals is a problem, then we should move the `MOD_ID` constant to a more public location as thats the only reason im using `NeoForgeVersion` here for the mod id constant. 

I have also seen alot of people struggle to find is/think we dont have one/if they do find the constant question why its in the version class.
Clearly something people want to use and having it inside internals doesnt make sense, it should be part of the public api imo.

Technici4n replied:
It's fine to keep in the test - just prefer not to use this pattern in your mods. We can add a public API in a separate PR.

Technici4n reviewed (changes requested):
I like this! I have a few minor comments.

ApexModder committed:
Prefer `optionalFieldOf` over `lenientOptionalFieldOf`

ApexModder committed:
Shorten generic type names

ApexModder committed:
Rename `ElementExistsCondition` to `RegisteredCondition`

Also rename `element` property to `value`

ApexModder committed:
Rename methods to better match the `RegisteredCondition` rename

`elementsExists` -> `isRegistered`
`itemExists` -> `itemRegistered`

ApexModder committed:
Convert `NeoForgeConditions` to be a utility class

Technici4n commented on src/main/java/net/neoforged/neoforge/common/conditions/NeoForgeConditions.java:
```diff
@@ -36,28 +36,28 @@ static ICondition or(ICondition... values) {
     }
 
-    static <TRegistry> ICondition elementExists(ResourceKey<TRegistry> registryKey) {
+    static <TRegistry> ICondition isRegistered(ResourceKey<TRegistry> registryKey) {
```
I would prefer just `registered`. That's also more consistent with `itemRegistered`.

ApexModder replied:
yeah tbh i couldnt decide which to use, went with `isRegistered` as it reads better imo, but like you say `registered` does match more with the other names.

ApexModder committed:
Rename `isRegistered` -> `registered`

ApexModder committed:
Rename `ELEMENT_EXISTS_CONDITION` -> `REGISTERED_CONDITION` in `NeoForgeMod`

left over from initial class rename
